File: scenes/InfiniteGallery/scene.ts
import Victor from "victor";
import * as THREE from "three";
import { events, consulters } from "scene-preset";
import { CanvasState } from "scene-preset/lib/types/state";
import { Scene, Scenes, SceneExport } from "scene-preset/lib/types/consulters";
import gsap from "gsap";

import Image from "../../meshes/Image";
import Text from "../../meshes/Text";
import Model from "../../meshes/Model";
import getTextureMaterial from "../../materials/getTextureMaterial";
import getQuixelMaterial from "../../materials/getQuixelMaterial";
import wavyMaterial from "../../materials/wavy";
import rainbowMaterial from "../../materials/rainbow";
import PointLightSet from "../../meshes/PointLightSet";

const scale = 4;
const distance = 5;
const height = 1;
const space = 3;

export default {
  pictures: {
    object: async () => {
      const hallway = new THREE.Group();

      //   const poems = await getImagesPromise({
      //     url: "https://luisarmando-testcoder.github.io/assets/poems/poem%20(###).png",
      //     max: 61,
      //     name: "poem"
      //   });

      //   hallway.add(...poems);

      const aiArtworks = await getImagesPromise({
        url: "https://luisarmando-testcoder.github.io/assets/ai-images/ai%20(###).png",
        max: 61,
        name: "ai-art",
      });

      hallway.add(...aiArtworks);

      return hallway;
    },
    onSetup({ object3D: hallway }: SceneExport) {
        const {length} = hallway.children

        hallway.children.forEach((picture, index) => {
            const opposite = Math.sign(index % 2 - .5)

            picture.translateX(distance * opposite)
            picture.rotateY(opposite * (Math.PI / 2))

            picture.position.z = (index - length / 2) * (scale + space)
            picture.position.y = height + scale / 2;
        })
    },
    onAnimation: ({ object3D: hallway }: SceneExport) => {
        const {length} = hallway.children

        hallway.children.forEach(({position}) => {
            position.z = ((position.z - .1) % ((length / 2) * (scale + space)))
        })
    },
  } as unknown as Scene,
} as Scenes;

async function getImagesPromise({
  url,
  min = 1,
  max,
  name,
}: {
  url: string;
  min?: number;
  max: number;
  name?: string;
}) {
  const imagesPromise: Promise<THREE.Group>[] = [];

  for (let index = min; index < max; index++) {
    const imgURL = url.replace(/###/g, String(index));
    const image = getImagePromise({ url: imgURL, scale, name });

    imagesPromise.push(image);
  }

  return Promise.all(imagesPromise);
}

async function getImagePromise({
  url,
  scale = 1,
  name,
}: {
  url: string;
  scale?: number;
  name?: string;
}): Promise<THREE.Group> {
  const textureLoader = new THREE.TextureLoader();
  const imageLoader = new THREE.ImageLoader();

  textureLoader.crossOrigin = "anonymous";
  imageLoader.crossOrigin = "anonymous";

  return new Promise((resolve) => {
    imageLoader.load(url, (image) => {
      const aspectRatio = image.width / image.height;
      const geometry = new THREE.PlaneGeometry(
        aspectRatio * scale,
        scale,
        1,
        1
      );
      const material = new THREE.MeshStandardMaterial({
        map: textureLoader.load(url),
        side: THREE.DoubleSide,
      });
      const imageObject = new THREE.Mesh(geometry, material);
      const resizedScale = scale * 1.02;
      const frameMaterial = new THREE.MeshStandardMaterial({
        side: THREE.DoubleSide,
        color: "#eee",
      });
      const frameGeometry = new THREE.PlaneGeometry(
        aspectRatio * resizedScale,
        resizedScale,
        1,
        1
      );
      const frame = new THREE.Mesh(frameGeometry, frameMaterial);

      frame.translateZ(0.01);

      const picture = new THREE.Group();

      picture.name = name || url;

      picture.add(frame);
      picture.add(imageObject);

      resolve(picture);
    });
  });
}

File: auto/scene.py
import os
import sys
from modules.main import executeConditionalPath, createFile

sceneName = input("Scene name: ")
folderPath = f"../scenes/{sceneName}"

executeConditionalPath(
    folderPath,
    lambda path: os.makedirs(path)
)

executeConditionalPath(
    f"{folderPath}/index.ts",
    lambda path: createFile(
        path,
        "import * as THREE from \"three\";\n" +
        "import presetScene, { consulters, types } from \"scene-preset\";\n" +
        "import scene from \"./scene\";\n" +
        "\n" +
        "let sceneEvents: {\n" +
        "  sceneGroup: THREE.Group;\n" +
        "  onSetup(canvasState: { [index: string]: any }): void;\n" +
        "  onAnimation(canvasState: { [index: string]: any }): void;\n" +
        "};\n" +
        "\n" +
        "export default (id: string) =>\n" +
        "  presetScene(\n" +
        "    {\n" +
        "      async setup(canvasState: { [index: string]: any }) {\n" +
        "        sceneEvents = await consulters.getSceneLifeCycle(scene);\n" +
        "\n" +
        "        sceneEvents?.onSetup(canvasState);\n" +
        "      },\n" +
        "      animate(canvasState: { [index: string]: any }) {\n" +
        "        sceneEvents?.onAnimation(canvasState);\n" +
        "      },\n" +
        "    },\n" +
        "    `#${id}`\n" +
        "  );\n"
    )
)

executeConditionalPath(
    f"{folderPath}/scene.ts",
    lambda path: createFile(
        path,
        "import Victor from \"victor\";\n" +
        "import * as THREE from \"three\";\n" +
        "import { events, consulters } from \"scene-preset\";\n" +
        "import { CanvasState } from \"scene-preset/lib/types/state\";\n" +
        "import { Scene, Scenes, SceneExport } from \"scene-preset/lib/types/consulters\";\n" +
        "import gsap from \"gsap\";\n" +
        "\n" +
        "import Image from \"../../meshes/Image\";\n" +
        "import Text from \"../../meshes/Text\";\n" +
        "import Model from \"../../meshes/Model\";\n" +
        "import getTextureMaterial from \"../../materials/getTextureMaterial\";\n" +
        "import getQuixelMaterial from \"../../materials/getQuixelMaterial\";\n" +
        "import wavyMaterial from \"../../materials/wavy\";\n" +
        "import rainbowMaterial from \"../../materials/rainbow\";\n" 
        "import PointLightSet from \"../../meshes/PointLightSet\";\n" +
        "\n" +
        "export default {\n" +
        "  \n" +
        "} as Scenes;\n"
    )
)

executeConditionalPath(
    f"../pages/{sceneName.lower()}.tsx",
    lambda path: createFile(
        path,
        "import Head from 'next/head'\n" +
        "import * as Components from '../components'\n" +
        "\n" +
        f"function {sceneName}() " + "{\n" +
        "    return (\n" +
        "        <div>\n" +
        "            <Head>\n" +
        f"                <title>{sceneName} | Home for 3D experiments</title>\n" +
        f"                <meta name=\"description\" content=\"{sceneName}\" />\n" +
        "                <link rel=\"icon\" href=\"/favicon.ico\" />\n" +
        "            </Head>\n" +
        "            <Components.L1.Canvas3D\n" +
        "                id={'" + sceneName + "'}\n" +
        "                scenes={[\n" +
        f"                    'Default',\n" +
        f"                    '{sceneName}',\n" +
        "                ]}\n" +
        "            />\n" +
        "        </div>\n" +
        "    )\n" +
        "}\n\n" +
        f"export default {sceneName};"
    )
)

import modules.export